### ISSUE TITLE
Failure to install the package on Windows

### ISSUE THREAD

**Issue by alex-hhh** [97jaz/tzinfo#9]
If I try to install this package on a Windows machine, using "raco pkg install tzinfo", the installation fails. This happens using both Racket 7.4 and Racket 7.4 CS:

```
collection-file-path: collection not found
  collection: "tzinfo\\private"
  in collection directories:
   C:\program files\racket\collects
   ... [170 additional linked and package directories]
```

It seems that, if I comment out the path list creation in "zoneinfo.rkt", the problem goes away and the package seems to work fine.  So the "fix" for me is to do the following:

```racket
(define-runtime-path-list tzdata-paths
  (match (find-relevant-directories '(tzdata-zoneinfo-dir))
    [(cons dir _)
     (define relpath ((get-info/full dir) 'tzdata-zoneinfo-dir))
     (define zoneinfo-dir (build-path dir relpath))

     (current-zoneinfo-search-path (list zoneinfo-dir))
     
     ;; (parameterize ([current-directory zoneinfo-dir])
     ;;  (for/list ([f (in-directory)])
     ;;     (list 'lib (path->string (build-path "tzinfo" relpath (path->string f))))))
     '()]
    [_ '()]))
```

I am not sure what `tzdata-paths` is used for, so I don't think this is the right fix.

**Comment by 97jaz** [comment]
@alex-hhh `tzdata-paths` is really there to register the paths of the data files provided by the `tzdata` package so that if you compile a standalone program, the compiler will know to package those with the program.

I haven't seen this failure before, but I also don't own a Windows machine. I do test Windows-specific changes on a VM, so I'm surprised by this. It could be the result of a somewhat recent change in Racket, or it could be something else entirely. I can do another manual test in a VM, and maybe I can set up a Windows build in CI.

Thanks for letting me know.

**Comment by 97jaz** [comment]
Oh also, what version of Windows are you running?

**Comment by alex-hhh** [comment]
I am running windows 10, but I don't think the windows version is the problem.  

I did some more investigations and I think the problem is that `build-path` constructs paths with windows style separators on windows -- I printed them out and they look like this: "tzinfo\private/data\Australia\Perth".  I think the `'lib` module definition type does not like backslashes as separators and using a string-replace fixed the problem (see code snipped below).

I think `build-paths` (or one of its variants) could be used to produce unix style paths on all platforms, but the workings of that is complicated (there is some threads about that on `racket-users`), so I think using `string-replace` is the simplest option, but you might have a different opinion.

```racket
(define-runtime-path-list tzdata-paths
  (match (find-relevant-directories '(tzdata-zoneinfo-dir))
    [(cons dir _)
     (define relpath ((get-info/full dir) 'tzdata-zoneinfo-dir))
     (define zoneinfo-dir (build-path dir relpath))

     (current-zoneinfo-search-path (list zoneinfo-dir))

     (define (make-lib-path f)
       (define path (path->string (build-path "tzinfo" relpath (path->string f))))
       (string-replace path "\\" "/"))
     
     (parameterize ([current-directory zoneinfo-dir])
       (for/list ([f (in-directory)])
         (list 'lib (make-lib-path f))))]
    [_ '()]))
```

**Comment by alex-hhh** [comment]
I have also verified that with my fix above, the files are correctly copied into a stand-alone distribution when "raco dist" is used.

**Comment by 97jaz** [comment]
I think I was just overcomplicating things by using `'lib` paths. How about:
```racket
(define-runtime-path-list tzdata-paths
  (match (find-relevant-directories '(tzdata-zoneinfo-dir))
    [(cons dir _)
     (define relpath ((get-info/full dir) 'tzdata-zoneinfo-dir))
     (define zoneinfo-dir (build-path dir relpath))

     (current-zoneinfo-search-path (list zoneinfo-dir))
     
     (parameterize ([current-directory zoneinfo-dir])
       (for/list ([f (in-directory)])
         (build-path zoneinfo-dir f)))]
    [_ '()]))
```

**Comment by alex-hhh** [comment]
Your proposed solution has a subtle bug: `define-runtime-path-list` will now receive a full path, not a relative one, and this means that all path references will be to the original place where the tzdata package was installed.  The files will still be copied into the distribution by "raco dist", but if you remove the package, or move the distribution to another machine without Racket installed, it will fail to run. 

Just to be sure, I tested that this is the case with the code you proposed, but I have been caught by this subtlety of `define-runtime-path` before :-)  I think "raco dist" should give out a warning for this case, but I don't understand the full implications.

**Comment by 97jaz** [comment]
Makes sense -- the `'lib` paths are more resilient. I'm still looking for something a bit more robust, though. A path on a UNIX system might actually have a `\` in the filename (not a good idea, but possible). I'm thinking maybe:
```racket
(define-runtime-path-list tzdata-paths
  (match (find-relevant-directories '(tzdata-zoneinfo-dir))
    [(cons dir _)
     (define relpath ((get-info/full dir) 'tzdata-zoneinfo-dir))
     (define zoneinfo-dir (build-path dir relpath))

     (current-zoneinfo-search-path (list zoneinfo-dir))

     (define (make-lib-path f)
       (define segments
         (list* "tzinfo"
                relpath
                (map path->string (explode-path f))))

       `(lib ,(string-join segments "/")))
     
     (parameterize ([current-directory zoneinfo-dir])
       (for/list ([f (in-directory)])
         (make-lib-path f)))]
    [_ '()]))
```

**Comment by alex-hhh** [comment]
The above code also does not work when a distribution is created than the package removed.  In fact my original proposal does not work either in this case.  

From what I can tell, the search path is set by `current-zoneinfo-search-path` to `zoneinfo-dir` and this is an absolute path.  While `define-runtime-paths` will package the files, the code will still look in the original place.  I think the code will need to change to use the `tzdata-paths` variable.

**Comment by 97jaz** [comment]
Ah, good point. We'll need a runtime path specifically for the root of the zoneinfo db.

**Comment by alex-hhh** [comment]
I spent some more time trying to diagnose this issue, and, while I now have a better understanding of it, I am not sure I know of a good solution.  

In any case, the problem is that the body of the `define-runtime-paths` definition is always evaluated, even when running an executable created with "raco dist".  For example, I defined the `tzdata-search-path` variable as below:

```racket
(define-runtime-path-list tzdata-search-path
  (match (find-relevant-directories '(tzdata-zoneinfo-dir))
    [(cons dir _)
     (define relpath ((get-info/full dir) 'tzdata-zoneinfo-dir))
     (define zoneinfo-dir (build-path dir relpath))
     (list `(lib ,(string-join (list "tzinfo" relpath) "/")))]
    [_  '()]))
```

Even at runtime, in a standalone distribution, created with "raco dist" the above code will:

* call `find-relevant-directories` and find a path in the `tzdata` package in the racket distribution
* the `(cons dir _)` case is matched in the `match` statement
* retrieve the `tzdata-zoneinfo-dir` key, which is the string "private/data"
* construct the path as `(lib "tzinfo/private/data")`
* this path is than adjusted at runtime to point to a path inside the distribution directory of the stand-alone distribution.

However, the first step in this process, `find-relevant-directories`, requires a Racket installation.  If I remove the installation (for this test I just renamed my local package directory so Racket cannot find it), the new steps become:

* call `find-relevant-directories` and find a path in the `tzdata` package in the racket distribution, this returns an empty list, since the package is not found anymore
* the "default" entry in the `match` statement is now matched,
* this causes `tzdata-search-path` to be set to the empty list and the timezone files are not found even though they are packaged in the distribution.

The solution that I can see is to avoid fetching the tzdata paths from the package info file, and just use the "tzinfo/private/data" directly.  The code below works correctly -- it looks like "raco dist" can copy data recursively, so there is no need to explicitly list each directory in `tzdata-paths`

```racket
(define-runtime-path-list tzdata-search-path
  (if (eq? (system-type) 'windows)
      `((lib "tzinfo/private/data"))
      '()))

(unless (null? tzdata-search-path)
  (current-zoneinfo-search-path tzdata-search-path))
```

What do you think?

**Comment by 97jaz** [comment]
Fixed by #13 and #14

**Comment by alex-hhh** [comment]
I can confirm that the two pull requests have fixed this issue for me.  I can build and create a distribution of my application on Windows.  Thanks, Alex.
